File: advent_of_code/2021/2/solution.py
"""
Solution for AoC 2021 Day 2
https://adventofcode.com/2021/day/2
"""
from collections import namedtuple
from typing import List


Instruction = namedtuple("Instruction", ["direction", "distance"])
Coordinates = namedtuple("Coordinates", ["horizontal", "depth", "aim"])


def execute_instructions(instructions: List[Instruction]) -> Coordinates:
    location = Coordinates(0, 0, 0)

    for instruction in instructions:
        if instruction.direction == "forward":
            location = Coordinates(
                location.horizontal + instruction.distance, location.depth, location.aim
            )
        elif instruction.direction == "up":
            location = Coordinates(
                location.horizontal, location.depth - instruction.distance, location.aim
            )
        else:  # instruction.direction == "down"
            location = Coordinates(
                location.horizontal, location.depth + instruction.distance, location.aim
            )

    return location


def execute_instructions_with_aim(instructions: List[Instruction]) -> Coordinates:
    location = Coordinates(0, 0, 0)

    for instruction in instructions:
        if instruction.direction == "forward":
            location = Coordinates(
                location.horizontal + instruction.distance,
                location.depth + (instruction.distance * location.aim),
                location.aim,
            )
        elif instruction.direction == "up":
            location = Coordinates(
                location.horizontal, location.depth, location.aim - instruction.distance
            )
        else:  # instruction.direction == "down"
            location = Coordinates(
                location.horizontal, location.depth, location.aim + instruction.distance
            )

    return location


def multiply_coordinates(coordinates: Coordinates) -> int:
    return coordinates.depth * coordinates.horizontal


def main():
    def parse_instruction_from_text(line: str) -> Instruction:
        direction, distance = line.split(" ")
        return Instruction(direction, int(distance))

    with open("input.txt") as handle:
        instructions = [parse_instruction_from_text(line) for line in handle]

    print(f"Step 1: {multiply_coordinates(execute_instructions(instructions))}")
    print(
        f"Step 2: {multiply_coordinates(execute_instructions_with_aim(instructions))}"
    )


if __name__ == "__main__":
    main()

File: advent_of_code/2021/2/solution.go
package main

// Solution for AoC 2021 Day 2
// https://adventofcode.com/2021/day/2

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type instruction struct {
	direction string
	distance  int
}

func ReadInstructions(r io.Reader) ([]instruction, error) {
	scanner := bufio.NewScanner(r)
	scanner.Split(bufio.ScanLines)

	var instructions []instruction

	for scanner.Scan() {
		split := strings.Split(scanner.Text(), " ")
		var instruction instruction
		instruction.direction = split[0]
		val, err := strconv.Atoi(split[1])
		if err != nil {
			return nil, err
		}
		instruction.distance = val
		instructions = append(instructions, instruction)
	}

	return instructions, scanner.Err()
}

type coordinate struct {
	horizontal int
	depth      int
	aim        int
}

func ExecuteInstructions(instructions []instruction) coordinate {
	location := coordinate{0, 0, 0}

	for _, instruction := range instructions {
		switch instruction.direction {
		case "forward":
			location.horizontal += instruction.distance
		case "up":
			location.depth -= instruction.distance
		case "down":
			location.depth += instruction.distance
		}
	}

	return location
}

func ExecuteInstructionsWithAim(instructions []instruction) coordinate {
	location := coordinate{0, 0, 0}

	for _, instruction := range instructions {
		switch instruction.direction {
		case "forward":
			location.horizontal += instruction.distance
			location.depth += (location.aim * instruction.distance)
		case "up":
			location.aim -= instruction.distance
		case "down":
			location.aim += instruction.distance
		}
	}

	return location
}

func MultiplyCoordinates(coordinate coordinate) int {
	return coordinate.horizontal * coordinate.depth
}

func main() {
	handle, err := os.Open("input.txt")
	check(err)

	defer handle.Close()

	instructions, err := ReadInstructions(handle)
	check(err)

	fmt.Printf("Part 1: %d\n", MultiplyCoordinates(ExecuteInstructions(instructions)))
	fmt.Printf("Part 2: %d\n", MultiplyCoordinates(ExecuteInstructionsWithAim(instructions)))
}
